### PR TITLE
MBS-9086 Basic AGP 4.1 support

### PR DESCRIPTION
**Pull request by eugene-krivobokov**
- Overview: https://medium.com/androiddevelopers/new-apis-in-the-android-gradle-plugin-f5325742e614
- Samples: https://github.com/android/gradle-recipes

New lifecycle in AGP tasks:

- Build scripts are run, allowing the build and plugins to configure the Android Gradle Plugin DSL objects as well as registering variant callbacks.   
- The Android Gradle Plugin combines build types and product flavors to create variants and test components.   
- The `onVariants` API is invoked for each variant, allowing the customization of variant settings.
     - NB: It doesn’t use properties since the values are used at configuration time.   
- The `onVariantProperties` API is invoked for each variant that was enabled allowing the registration of tasks that consume or modify intermediates of the build.
     - NB: Didn't find alternative for test variants. Asking for help: https://gradle-community.slack.com/archives/CJYS1DAP5/p1605196130059800
- The Android Gradle Plugin registers tasks for the variants after both the `onVariants` and the `onVariantProperties` callbacks have been called.   
- The previous variant API (`applicationVariants`, `testVariants` and so on) is called for each variant, using the registered tasks.
     - NB: These abstractions changed a bit.  
- Gradle calculates the task graph, and the build can begin executing.

---
- [x] check integration

### CODE REVIEW

File: subprojects/gradle/android/src/main/kotlin/com/avito/android/PackageAndroidArtifactExtensions.kt [hunk 1]
```diff
@@ -1,14 +1,27 @@
...
 import com.android.build.gradle.tasks.PackageAndroidArtifact
+import org.gradle.api.file.Directory
+import org.gradle.api.file.DirectoryProperty
 import java.io.File
 
-fun PackageAndroidArtifact.getApkFile(): File {
-    require(apkNames.size == 1) {
-        "Cannot get apk from $name android artifact task because apkNames.size != 1 ($apkNames). Split apk is not supported."
-    }
+// TODO: Use Artifacts API
```
Comment (eugene-krivobokov) [inline subprojects/gradle/android/src/main/kotlin/com/avito/android/PackageAndroidArtifactExtensions.kt:8]: Didn't find a producer of test APK (`onTestComponentProperties` or similar API)

File: subprojects/gradle/android/src/main/kotlin/com/avito/android/ProjectExtensions.kt [hunk 2]
```diff
@@ -26,6 +29,10 @@ fun Project.withAndroidModule(block: (testedExtension: TestedExtension) -> Unit)
     withAndroidLib(block)
 }
 
+@Suppress("UnstableApiUsage")
```
Comment (eugene-krivobokov) [inline subprojects/gradle/android/src/main/kotlin/com/avito/android/ProjectExtensions.kt:32]: Also, where are loads of TODOs in sources
I propose to use this new API only when we don't have an alternative.

File: subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/FakeInstrumentationTestsActionParams.kt [hunk 3]
```diff
@@ -15,8 +15,6 @@ import java.io.File
 fun InstrumentationTestsAction.Params.Companion.createStubInstance(
     mainApk: File = File(""),
     testApk: File = File(""),
-    apkOnTargetCommit: File = File(""),
-    testApkOnTargetCommit: File = File(""),
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/FakeInstrumentationTestsActionParams.kt:19]: Unused properties

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/InstrumentationTestsPlugin.kt [hunk 4]
```diff
@@ -189,15 +189,15 @@ class InstrumentationTestsPlugin : Plugin<Project> {
 
                                 if (extensionData.applicationApk == null) {
                                     task.dependencyOn(testedVariantPackageTask) { dependentTask ->
-                                        task.application.set(dependentTask.getApkFile())
+                                        task.application.set(dependentTask.apkDirectory())
```
Comment (eugene-krivobokov) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/InstrumentationTestsPlugin.kt:192]: https://github.com/avito-tech/avito-android/pull/632/files#r522296710

File: subprojects/gradle/signer/src/main/kotlin/com/avito/plugin/SignTask.kt [hunk 5]
```diff
@@ -42,11 +58,36 @@ abstract class SignTask @Inject constructor(objects: ObjectFactory) : DefaultTas
...
+
+    /**
+     * TODO: find a better way to profile a final artifact to CI
+     * Results of transformations are stored in build/intermediates/bundle/release/<task name>/out
+     * It's accessible by Artifacts API programmatically but we need a final one file.
+     * We rewrite an original file to preserve legacy behaviour.
+     */
+    private fun hackForArtifactsApi(inputFile: File, outputFile: File) {
+        outputFile.copyTo(inputFile, overwrite = true)
+    }
```
Comment (eugene-krivobokov) [inline subprojects/gradle/signer/src/main/kotlin/com/avito/plugin/SignTask.kt]: It works as before but seems incorrect by design :hankey:

**Review by dsvoronin — APPROVED** [review]

**Pull request merged**